### DISCUSSION TITLE
How to forward focus to conditional subcomponent?

### DISCUSSION THREAD

**Discussion by Argent77** [slint-ui/slint#2777] [Q&A] [answered]
Hi, a LineEdit component on a subcomponent should receive the focus as soon as the window is shown. However, I can't reference the subcomponent, since it is shown conditionally. Is there a way to solve or work around this issue?

<details>

<summary>Slint code (<a href="https://slint-ui.com/releases/1.0.2/editor/?snippet=import+%7B+Button%2C+LineEdit+%7D+from+%22std-widgets.slint%22%3B%0A%0Aexport+component+SubPage1+inherits+VerticalLayout+%7B%0A++++padding%3A+8px%3B%0A++++forward-focus%3A+input%3B%0A%0A++++input+%3A%3D+LineEdit+%7B%0A++++++++placeholder-text%3A+%22Enter+text+here%22%3B%0A++++%7D%0A%7D%0A%0Aexport+component+SubPage2+inherits+VerticalLayout+%7B%0A++++padding%3A+8px%3B%0A%0A++++button+%3A%3D+Button+%7B%0A++++++++text%3A+%22Click+me%22%3B%0A++++%7D%0A%7D%0A%0Aexport+component+Main+inherits+Window+%7B%0A++++private+property+%3Cint%3E+page-index%3A+0%3B%0A%0A++++preferred-width%3A+200px%3B%0A++++preferred-height%3A+50px%3B%0A%0A++++%2F%2F+forward-focus%3A+sub-page1%3B+++++++%2F%2F+does+not+work%0A%0A++++if+%28page-index+%3D%3D+0%29%3A+sub-page1+%3A%3D+SubPage1+%7B%7D%0A++++if+%28page-index+%21%3D+0%29%3A+sub-page2+%3A%3D+SubPage2+%7B%7D%0A%7D">SlintPad</a>)</summary>

```js
import { Button, LineEdit } from "std-widgets.slint";

export component SubPage1 inherits VerticalLayout {
    padding: 8px;
    forward-focus: input;

    input := LineEdit {
        placeholder-text: "Enter text here";
    }
}

export component SubPage2 inherits VerticalLayout {
    padding: 8px;

    button := Button {
        text: "Click me";
    }
}

export component Main inherits Window {
    private property <int> page-index: 0;

    preferred-width: 200px;
    preferred-height: 50px;

    // forward-focus: sub-page1;       // does not work

    if (page-index == 0): sub-page1 := SubPage1 {}
    if (page-index != 0): sub-page2 := SubPage2 {}
}
```

</details>

**Comment by tronical** [comment, marked as answer]
One thing that you could do is set the focus when `SubPage1` is created for the very first time, like [this](https://slint-ui.com/releases/1.0.2/editor/?snippet=import+%7B+Button%2C+LineEdit+%7D+from+%22std-widgets.slint%22%3B%0A%0Aexport+component+SubPage1+inherits+VerticalLayout+%7B%0A++++padding%3A+8px%3B%0A++++forward-focus%3A+input%3B%0A%0A++++input+%3A%3D+LineEdit+%7B%0A++++++++placeholder-text%3A+%22Enter+text+here%22%3B%0A++++%7D%0A%7D%0A%0Aexport+component+SubPage2+inherits+VerticalLayout+%7B%0A++++padding%3A+8px%3B%0A%0A++++button+%3A%3D+Button+%7B%0A++++++++text%3A+%22Click+me%22%3B%0A++++%7D%0A%7D%0A%0Aexport+component+Main+inherits+Window+%7B%0A++++private+property+%3Cint%3E+page-index%3A+0%3B%0A%0A++++preferred-width%3A+200px%3B%0A++++preferred-height%3A+50px%3B%0A%0A++++%2F%2F+forward-focus%3A+sub-page1%3B+++++++%2F%2F+does+not+work%0A++++property+%3Cbool%3E+first-time%3A+true%3B%0A%0A++++if+%28page-index+%3D%3D+0%29%3A+sub-page1+%3A%3D+SubPage1+%7B%0A++++++++init+%3D%3E+%7B%0A++++++++++++if+%28parent.first-time%29+%7B%0A++++++++++++++++self.focus%28%29%3B%0A++++++++++++%7D%0A++++++++++++parent.first-time+%3D+false%3B%0A++++++++%7D%0A++++%7D%0A++++if+%28page-index+%21%3D+0%29%3A+sub-page2+%3A%3D+SubPage2+%7B%7D%0A%7D): In the `init` callback, call `self.focus()`.

Careful, if you try this in SlintPad it's going to steal the focus - it can be annoying when editing :-)

**Comment by Argent77** [comment]
Yes, that works for me. Thank you.
